text/0314-cdk-construct-hub.md

```diff
@@ -362,6 +367,96 @@ pipeline.
+This will ensure that breaking changes will manifest as compile time errors, and force us to handle them.
+
+To that end, we create a dedicated package, called `construct-hub-docgen`, which will define this schema, and provide an API to produce it.
+The version of the schema file will be derived from the major version of the package.
+Every time a change is made to the schema, we automatically detect what type of change has been made. (i.e breaking or not)
+
+> This can be done by using either [json-schema-diff](https://www.npmjs.com/package/json-schema-diff), or [jsii-diff](https://github.com/aws/jsii/tree/main/packages/jsii-diff).
+
+If a breaking change is detected, we enforce that the next version to be released will include a major version bump.
+(i.e the commit was accompanied with a *BREAKING CHANGE:* notice)
+
+> This can be done by storing the version prior to the bump, and comparing with the one after.
+
+When a new major version of the schema is released, it is automatically picked up and
+deployed to the backend. Since the schema is overidden, this will break the front-end
+because the version supporting the new schema hasn't been deployed yet.
+We cannot afford this, which means we must enforce a deployment order in this scenario,
+by which the backend is deployed **only after** the front-end supporting it is deployed.
+
+We can enforce this by employing the following mechanism:
+
+- The front-end declares a `peerDepenedency` on `construct-hub-docgen`. (e.g `^1.0.0`)
+- The backend declares a `devDependency` on `construct-hub-docgen` (e.g `1.0.0`)
+
+When a new version of `construct-hub-docgen` is released (e.g `2.0.0`):
+
+- If the backend picks up this version first, the resolved version in the closure
+will be `2.0.0`. This will conflict with the version that the front-end requires,
+and thus fail at install time.
+- If the front-end picks up this version first, it will be updated and a new version
+will be released that depends on `2.0.0`. At this point, when backend dependencies
+are upgraded, both `construct-hub-docgen`, and the front-end itself are upgraded,
+which is the desired outcome.
+
+> In addition, so that don't rely on `npm` behavior to reject this change,
+we can write a unit test that compares the schema version being used by the front-end,
+to the one being used by the backend, they have to match.
+
+This mechanism enforces that when the backend picks up a new version major of the schema,
+it will have to pick up a new version of the front-end that supports it.
+
+From this point on, we just need to make sure we deploy the front-end first in the CDK application.
+
+In order for the front-end to support multiple versions of the schema, it will have to:
+
+1. Fetch the schema from `/../docs-{lang}.json`
+2. Extract the version field.
+3. Based on the version, render the appropriate component.
```

Comment:
This seems to imply the front-end will need to have the type definitions of both the previous and current versions at the same time.

This means a whole section is missing that would describe what to do when a breaking change happens on the package that contains the type definitions/schema - namely the `next` version needs to also contain the type for the previous major... This can be achieved using a discriminated type union (those are neat):

```ts
export type PackageDescription = PackageDescriptionV1 | PackageDescriptionV2;

export const enum SchemaVersion {
  V1 = 'v1',
  V2 = 'v2',
}

export interface PackageDescriptionV1 {
  readonly version: SchemaVersion.V1;

  // Other properties
}

export interface PackageDescriptionV2 {
  readonly version: SchemaVersion.V2;

  // Other properties
}
```

Reply:
> This seems to imply the front-end will need to have the type definitions of both the previous and current versions at the same time.

I kind of considered this an implementation detail we'll sort out later. 

Since the front-end will actively maintain only the latest major version, we could naively switch to `any` in the components that use an older major version (this code is supposed to locked, untouched once a new major version is introduced). This way we don't need to keep older schema definitions in the package, just the latest one.

Going with your approach, slightly deviated, we could also have each component version import the appropriate schema definition, exposed from the schema package:

```ts
import { spec } from 'jsii-docgen/spec/v1`

export class PackageDocsV1 {
}
```

```ts
import { spec } from 'jsii-docgen/spec/v2`

export class PackageDocsV2 {
}
```

Btw, with this approach, the backend can have full control over when to start generating new schemas.

Do you feel have to flush this out right now? I'd like to get the implementation going and than update the RFC with the approach we decide on.

Reply:
I think the doc should probably mention there should be a provision around what will happen when we have to major-bump the schema, but I guess we can defer determining what the actual process is to a later point (when we actually need that - since we should be striving to avoid resorting to this as much as possible).

Reply:
> I think the doc should probably mention there should be a provision around what will happen when we have to major-bump the schema

Right now this is the provision :)

*Based on the version, render the appropriate component.*

I'll defer the actual process to a later point.